### PR TITLE
Migrate most of `rustc_builtin_macros` to diagnostic impls

### PR DESCRIPTION
**Pull request by clubby789**
cc #100717

This is a couple of days work, but I decided to stop for now before the PR becomes too big. There's around 50 unresolved failures when `rustc::untranslatable_diagnostic` is denied, which I'll finish addressing once this PR goes thtough

A couple of outputs have changed, but in all instances I think the changes are an improvement/are more consistent with other diagnostics (although I'm happy to revert any which seem worse)

### CODE REVIEW

File: compiler/rustc_builtin_macros/messages.ftl [hunk 1]
```diff
@@ -3,3 +3,146 @@ builtin_macros_requires_cfg_pattern =
...
+
+builtin_macros_cfg_accessible_indeterminate = cannot determine whether the path is accessible or not
+
+builtin_macros_concat_bytestr = cannot concatenate a byte string literal
+
+builtin_macros_concat_missing_literal = expected a literal
+    .note = only literals (like `"foo"`, `-42` and `3.14`) can be passed to `concat!()`
+
+builtin_macros_concat_bytes_missing_literal = expected a byte literal
+    .note = only byte literals (like `b"foo"`, `b's'` and `[3, 4, 5`) can be passed to `concat_bytes!()`
```
Comment (fbstj) [inline compiler/rustc_builtin_macros/messages.ftl:31]:
```suggestion
    .note = only byte literals (like `b"foo"`, `b's'` and `[3, 4, 5]`) can be passed to `concat_bytes!()`
```

File: compiler/rustc_builtin_macros/src/alloc_error_handler.rs [hunk 2]
```diff
@@ -31,7 +32,7 @@ pub fn expand(
         {
             (item, true, ecx.with_def_site_ctxt(fn_kind.sig.span))
         } else {
-            ecx.sess.parse_sess.span_diagnostic.span_err(item.span(), "alloc_error_handler must be a function");
+            ecx.sess.parse_sess.span_diagnostic.emit_err(errors::AllocErrorMustBeFn {span: item.span()});
```
Comment (compiler-errors) [inline compiler/rustc_builtin_macros/src/alloc_error_handler.rs:35]: format nit

```suggestion
            ecx.sess.parse_sess.span_diagnostic.emit_err(errors::AllocErrorMustBeFn {span: item.span() });
```

File: compiler/rustc_builtin_macros/src/concat.rs [hunk 3]
```diff
@@ -31,7 +33,7 @@ pub fn expand_concat(
                     accumulator.push_str(&b.to_string());
                 }
                 Ok(ast::LitKind::Byte(..) | ast::LitKind::ByteStr(..)) => {
-                    cx.span_err(e.span, "cannot concatenate a byte string literal");
+                    cx.emit_err(errors::ConcatBytestr {span: e.span});
```
Comment (compiler-errors) [inline compiler/rustc_builtin_macros/src/concat.rs:36]:
```suggestion
                    cx.emit_err(errors::ConcatBytestr { span: e.span });
```

File: compiler/rustc_builtin_macros/src/concat.rs [hunk 4]
```diff
@@ -55,7 +57,7 @@ pub fn expand_concat(
                 }
             }
             ast::ExprKind::IncludedBytes(..) => {
-                cx.span_err(e.span, "cannot concatenate a byte string literal")
+                cx.emit_err(errors::ConcatBytestr {span: e.span});
```
Comment (compiler-errors) [inline compiler/rustc_builtin_macros/src/concat.rs:60]:
```suggestion
                cx.emit_err(errors::ConcatBytestr { span: e.span });
```

File: compiler/rustc_builtin_macros/src/concat_bytes.rs [hunk 5]
```diff
@@ -167,7 +146,7 @@ pub fn expand_concat_bytes(
                         }
                     }
                 } else {
-                    cx.span_err(count.value.span, "repeat count is not a positive number");
+                    cx.emit_err(errors::ConcatBytesBadRepeat {span: count.value.span});
```
Comment (compiler-errors) [inline compiler/rustc_builtin_macros/src/concat_bytes.rs:149]: same here

File: compiler/rustc_builtin_macros/src/format.rs [hunk 6]
```diff
@@ -557,6 +543,7 @@ fn invalid_placeholder_type_error(
             ("x", "LowerHex"),
             ("X", "UpperHex"),
         ] {
+            // FIXME: allow derive API to create tool-only suggestions
```
Comment (compiler-errors) [inline compiler/rustc_builtin_macros/src/format.rs:546]: `style = "tool-only",` ?

**Review by compiler-errors — APPROVED** [review]
one comment

File: compiler/rustc_builtin_macros/src/format.rs (hunk 6 above)
Comment (clubby789) [reply compiler/rustc_builtin_macros/src/format.rs:546]: Oops, thank you

File: tests/ui/macros/macros-nonfatal-errors.stderr [hunk 7]
```diff
@@ -103,8 +115,14 @@ help: try removing these
    |
 LL |     #[default]
    |     ^^^^^^^^^^
+help: try removing these
```
Comment (compiler-errors) [inline tests/ui/macros/macros-nonfatal-errors.stderr:118]: Why is this rendering as several helps rather than one multispan?

Comment (clubby789) [reply tests/ui/macros/macros-nonfatal-errors.stderr]: Updated the derives to allow `#[help]` on MultiSpan

**Comment by compiler-errors** [comment]
ty

@bors delegate+

r=me when green

**Pull request merged**